Feat: Improve the precision of label bounding box calculations and rendering

fzaman-linz opened the pull request:
This PR improves the precision of label bounding box calculations and rendering by introducing new style properties and refining label metrics handling.

- `text-metrics (font | glyph)`
Allows to choose between browser-default text metrics or actual font metrics (actualBoundingBoxAscent / Descent) for more accurate label height calculations.

- Label Dimension Calculation Enhancements
`calculateLabelDimensions()` now supports `text-metrics: glyph`, using `actualBoundingBoxAscent` and `actualBoundingBoxDescent` for more precise vertical alignment and height.

Associated issues: 

- #3417 



**Checklist**

Author:

- [x] The proper base branch has been selected.  New features go on `unstable`.  Bug-fix patches can go on either `unstable` or `master`.
- [x] Automated tests have been included in this pull request, if possible, for the new feature(s) or bug fix.  Check this box if tests are not pragmatically possible (e.g. rendering features could include screenshots or videos instead of automated tests).
- [x] The associated GitHub issues are included (above).
- [x] Notes have been included (above).
- [x] For new or updated API, the `index.d.ts` Typescript definition file has been appropriately updated.

Reviewers:

- [x] All automated checks are passing (green check next to latest commit).
- [x] At least one reviewer has signed off on the pull request.
- [x] For bug fixes:  Just after this pull request is merged, it should be applied to both the `master` branch and the `unstable` branch.  Normally, this just requires cherry-picking the corresponding merge commit from `master` to `unstable` -- or vice versa.

fzaman-linz commented:
Hi guys, could you please have a look

> cc: @maxkfranz @mikekucera

fzaman-linz commented:
Thanks for the feedback @mikekucera.
I've made the changes you suggested and here is a demo: https://codepen.io/fzaman-linz/pen/pvEYvxj
cc: @d2fong @maxkfranz

maxkfranz commented:
> <img alt="Screenshot 2026-02-16 at 2 26 10 PM" width="856" height="548" src="https://private-user-images.githubusercontent.com/5350528/550607754-92924a2d-77d5-4e63-8d95-c636191c99f3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY3MDY4NDAsIm5iZiI6MTc3NjcwNjU0MCwicGF0aCI6Ii81MzUwNTI4LzU1MDYwNzc1NC05MjkyNGEyZC03N2Q1LTRlNjMtOGQ5NS1jNjM2MTkxYzk5ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MjBUMTczNTQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODlmYjFiOGVjMzg5MGFlMWQ0ZTM2OGI3ZjMxZWM1YWUwMzc5MWYxMDVhMTgzODRmY2Q5NDNkMjNmNjJiYzllYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.1h3vsb16x5oxUDG93-bduJ6OumHC6B97TrmXbVU5QtQ">

What happens if you have a label like 'apple' instead of just 'a'?

If you have several nodes like the above image, the text (baseline) on all the labels should line up.  That's important.

Try these labels.  They should all line up as long as the label positions are the same and the node dimensions are the same:

- 'yes' (descender)
- 'no' (no ascender nor descender)
- 'not' (ascender)

fzaman-linz commented:
@maxkfranz Yes, label text baselines won't line up across the board. Tight bounding boxes need per‑word heights (which shifts baselines), while consistent baselines need fixed heights (which reintroduces the gaps we're trying to avoid). That's why keeping `text-metrics: default | actual` makes sense so it leaves the choice with the user.
Are you ok with that?
<img width="597" height="133" alt="image" src="https://github.com/user-attachments/assets/98a52b9b-83aa-4520-9c71-8d64ce149964" />

maxkfranz commented:
@fzaman-linz, the team will discuss this on a call on Thursday 23 April.

maxkfranz commented:
@fzaman-linz, in the meantime, would you please update the OP with the current status of the feature (i.e. only currently proposed props)?

fzaman-linz committed:
feat: text-metrics (default | actual)

fzaman-linz commented:
@maxkfranz I've updated the OP and also updated the example URL to showcase default vs actual.
https://codepen.io/fzaman-linz/pen/pvEYvxj
<img width="656" height="275" alt="image" src="https://github.com/user-attachments/assets/427983b4-9049-48b5-a922-08ce19ffca16" />

maxkfranz commented:
Great. Thanks

maxkfranz commented:
The team is discussing this feature and the details of the properties etc., and we'll have an update some time next week.

maxkfranz commented on src/collection/dimensions/bounds.mjs:
```diff
@@ -291,7 +291,8 @@ let updateBoundsFromLabel = function( bounds, ele, prefix ){
     let halfBorderWidth = borderWidth / 2;
     let padding = ele.pstyle( 'text-background-padding' ).pfValue;
-    let marginOfError = 2; // expand to work around browser dimension inaccuracies
+    let marginOfErrorX = ele.pstyle( 'bbox-margin-error-x' ).pfValue; // expand to work around browser dimension inaccuracies
```
Are larger values than 2 needed for specific cases?

maxkfranz replied:
From @mikekucera : 

> Remove the bbox-margin-error-x / bbox-margin-error-y properties, the default of 2 is fine. Its not worth it to add two VPs just for this.

fzaman-linz replied:
Fair, will remove bbox-margin-error-x/y, keep the hardcoded 2. The tradeoff isn't worth it. Once we decide on the `text-metrics` arguments name I'll update it.

maxkfranz commented on src/style/properties.mjs:
```diff
@@ -91,6 +91,7 @@ const styfn = {};
     valign: { enums: [ 'top', 'center', 'bottom' ] },
     halign: { enums: [ 'left', 'center', 'right' ] },
     justification: { enums: [ 'left', 'center', 'right', 'auto' ] },
+    textMetrics: { enums: [ 'default', 'actual' ] },
```
These values should be more descriptive so they're obvious as to what they do.  E.g. 'aligned' | 'tight'.  Or whatever names would be clear.

fzaman-linz replied:
How about text-metrics: `font` | `glyph`, or `auto` | `precise`? 
Default would be `font` / `auto`, but happy to go with whatever's preferred.

maxkfranz replied:
font | glyph sounds intuitive/clear

maxkfranz commented on src/extensions/renderer/base/coord-ele-math/labels.mjs:
```diff
@@ -312,14 +312,14 @@ BRp.applyPrefixedLabelDimensions = function( ele, prefix ){
 
   let labelDims = this.calculateLabelDimensions( ele, text );
   let lineHeight = ele.pstyle('line-height').pfValue;
+  let size = ele.pstyle('font-size').pfValue;
```
Could you upload some screenshots of some line-height:1 multiline labels?  We should verify that this change does not cause regressions.  There are likely edge cases that `normPerLineHeight` was handling.

fzaman-linz replied:
https://codepen.io/fzaman-linz/pen/pvEYvxj
<img width="356" height="373" alt="image" src="https://github.com/user-attachments/assets/ad2512f6-692c-459d-9fd7-c9d9ea682d77" />

maxkfranz replied:
Are the blue multiline cases comparable to the current prod version?

fzaman-linz replied:
Yes, they're comparable to the current prod version. Also I've made the suggested changes.

fzaman-linz committed:
feat: text-metrics (font | glyph)

slava-class commented:
One downstream datapoint that may be useful while testing this PR: we recently hit a separate initial-render label issue that looked like a Cytoscape label metrics bug, but the root cause was webfont readiness. Cytoscape was created before custom fonts had loaded, so the initial canvas label metrics/bounds were based on fallback fonts; a later style refresh made the labels correct.

For any demo/test cases here that use webfonts, it may be worth loading/checking the font before creating the graph. Preload was not sufficient in our case; we had to wait for `document.fonts.load(...)` plus `document.fonts.ready`, check the exact label font descriptors with `document.fonts.check(...)`, then instantiate Cytoscape. I added the concrete snippet on the directly related issue: https://github.com/cytoscape/cytoscape.js/issues/3408#issuecomment-4412270646
